### PR TITLE
Add Textfield for custom Pipeline/TaskRun name in PipelineRun/TaskRun create page. 

### PR DESCRIPTION
**Pull request by taechonkim**
# Changes
Add TextField to enter PipelineRun/TaskRun name (related https://github.com/tektoncd/dashboard/issues/2313)
In initial state, the default value (Pipeline or Taskname-Date.now() format) is entered. User can customize it if desired.

스크린샷 2022-04-23 오후 6 07 38

# Submitter Checklist

As the author of this PR, please check off the items in this checklist:

- [ ] [Docs](https://github.com/tektoncd/community/blob/main/standards.md#docs) included if any changes are user facing
- [ ] [Tests](https://github.com/tektoncd/community/blob/main/standards.md#tests) included if any functionality added or changed
- [x] Follows the [commit message standard](https://github.com/tektoncd/community/blob/main/standards.md#commits)
- [x] Meets the [Tekton contributor standards](https://github.com/tektoncd/community/blob/main/standards.md) (including
  functionality, content, code)

_See [the contribution guide](https://github.com/tektoncd/dashboard/blob/main/CONTRIBUTING.md)
for more details._

### CODE REVIEW

File: src/api/pipelineRuns.js [hunk 1]
```diff
@@ -78,6 +78,7 @@ export function deletePipelineRun({ name, namespace }) {
 export function createPipelineRun({
   namespace,
   pipelineName,
+  pipelineRunName,
```
Comment (AlanGreene) [inline src/api/pipelineRuns.js:81]: we should add the default value here so it's available to all consumers of the API
```diff
-  pipelineRunName,
+  name = `${pipelineName}-run-${Date.now()}`,
```

File: src/api/taskRuns.js [hunk 2]
```diff
@@ -78,6 +78,7 @@ export function cancelTaskRun({ name, namespace }) {
 export function createTaskRun({
   namespace,
   taskName,
+  taskRunName,
```
Comment (AlanGreene) [inline src/api/taskRuns.js:81]: same as for PipelineRun, we should have the default value in the API:
```diff
-  taskRunName,
+  name = `${taskName}-run-${Date.now()}`,
```

File: src/containers/CreatePipelineRun/CreatePipelineRun.js [hunk 3]
```diff
@@ -121,15 +123,18 @@ function CreatePipelineRun({ intl }) {
...
     },
     setState
   ] = useState({
     ...initialState,
     namespace: getNamespace(),
     pipelineRef: getPipelineName(),
     params: initialParamsState(null),
-    resources: initialResourcesState(null)
+    resources: initialResourcesState(null),
+    pipelineRunName: `${getPipelineName()}-run-${Date.now()}`
```
Comment (AlanGreene) [inline src/containers/CreatePipelineRun/CreatePipelineRun.js:137]: we should leave the initial state as an empty string, letting the API handle the default value so it's consistent anywhere it's used in future. This is also consistent with the other optional fields where they are left blank / unselected by default.

File: src/containers/CreatePipelineRun/CreatePipelineRun.js [hunk 4]
```diff
@@ -700,6 +714,23 @@ function CreatePipelineRun({ intl }) {
...
           />
+          <TextInput
+            id="create-pipelinerun--pipelinerunname"
+            labelText={intl.formatMessage({
+              id: 'dashboard.createRun.pipelineRunNameLabel',
+              defaultMessage: 'PipelineRun name'
+            })}
+            invalid={validationError && !validPipelineRunName}
+            invalidText={intl.formatMessage({
+              id: 'dashboard.createRun.invalidRunName',
```
Comment (AlanGreene) [inline src/containers/CreatePipelineRun/CreatePipelineRun.js:725]: Can we repurpose the existing string for this?

Rename `dashboard.createPipelineResource.nameError` to `dashboard.createResource.nameError` and use it here

**Review by AlanGreene — CHANGES_REQUESTED** [review]
Thanks for the PR @taechonkim, overall this looks great. I've made a few small suggestions, let me know if you have any questions.

/ok-to-test

File: src/api/pipelineRuns.js (hunk 1 above)
Comment (taechonkim) [reply src/api/pipelineRuns.js]: @AlanGreene Thank you :D I reflected your suggestion in https://github.com/tektoncd/dashboard/pull/2345/commits/5448a2166e592bdc15d097e4d29d3423463c2c0b

File: src/api/taskRuns.js (hunk 2 above)
Comment (taechonkim) [reply src/api/taskRuns.js]: Thanks for feedback :D reflected in https://github.com/tektoncd/dashboard/pull/2345/commits/b68845a1584c1128e3e31992e1e764d541615d34

**Comment by AlanGreene** [comment]
Hi @taechonkim, have you had a chance to look at the failures above? Let me know if you're having trouble resolving them.

There are two sets of errors:

- linting errors
  ```
   ERROR in /home/prow/go/src/github.com/tektoncd/dashboard/src/api/pipelineRuns.js
     103:45  error  'name' is already declared in the upper scope on line 81 column 3  no-shadow
     107:39  error  'name' is already declared in the upper scope on line 81 column 3  no-shadow

   /home/prow/go/src/github.com/tektoncd/dashboard/src/api/taskRuns.js
     114:51  error  'name' is already declared in the upper scope on line 81 column 3  no-shadow

   ✖ 3 problems (3 errors, 0 warnings)
   ```
- unexpected changes in the translation files, for example:
   ```diff
   diff --git a/src/nls/messages_zh-Hant.json b/src/nls/messages_zh-Hant.json
   index 1f8bb2cb..1e274f5b 100644
   --- a/src/nls/messages_zh-Hant.json
   +++ b/src/nls/messages_zh-Hant.json
   @@ -68,16 +68,16 @@
        "dashboard.createRun.invalidNamespace": "",
        "dashboard.createRun.invalidParams": "",
        "dashboard.createRun.invalidPipelineResources": "",
   -    "dashboard.createRun.invalidTimeout": "",
        "dashboard.createRun.invalidRunName": "",
   +    "dashboard.createRun.invalidTimeout": "",
        "dashboard.createRun.label.invalidText": "",
        "dashboard.createRun.labels.legendText": "",
        "dashboard.createRun.nodeSelector.legendText": "",
        "dashboard.createRun.optional.legendText": "",
   -    "dashboard.createRun.timeoutLabel": "",
   -    "dashboard.createRun.validationError": "",
        "dashboard.createRun.pipelineRunNameLabel": "",
        "dashboard.createRun.taskRunNameLabel": "",
   +    "dashboard.createRun.timeoutLabel": "",
   +    "dashboard.createRun.validationError": "",
        "dashboard.createTaskRun.createError": "",
        "dashboard.createTaskRun.errorLoading": "",
        "dashboard.createTaskRun.invalidTask": "",
   @@ -292,4 +292,4 @@
        "react-intl-formatted-duration.minutesUnit": "{value, plural, one {minute} other {minutes}}",
        "react-intl-formatted-duration.secondsUnit": "{value, plural, one {second} other {seconds}}"
      }
   -}
   +}
   \ No newline at end of file
   ```
   These are likely from manually editing these files instead of running the scripts to update them, e.g. `npm run i18n:extract`. Using the provided scripts ensures the correct line endings and encoding are used, as well as the correct ordering of the keys.

File: src/containers/CreatePipelineRun/CreatePipelineRun.js (hunk 3 above)
Comment (taechonkim) [reply src/containers/CreatePipelineRun/CreatePipelineRun.js]: @AlanGreene Thank you for your feedback. I reflected it at https://github.com/tektoncd/dashboard/pull/2345/commits/896d6821d3c6d7e4c12d3427a72b5a22ec691b2d

File: src/containers/CreateTaskRun/CreateTaskRun.js [hunk 5]
```diff
@@ -148,7 +152,8 @@ function CreateTaskRun({ intl }) {
     namespace: getNamespace(),
     taskRef: taskRefFromDetails,
     params: initialParamsState(null),
-    resources: initialResourcesState(null)
+    resources: initialResourcesState(null),
+    taskRunName: `${getTaskDetails().taskName}-run-${Date.now()}`
```
[the remark this answers is not in the document]
Comment (taechonkim) [reply src/containers/CreateTaskRun/CreateTaskRun.js]: @AlanGreene  Thanks for feedback. also reflected at https://github.com/tektoncd/dashboard/pull/2345/commits/896d6821d3c6d7e4c12d3427a72b5a22ec691b2d

**Comment by taechonkim** [comment]
@AlanGreene 
Sorry for the late response. 😢
I fixed errors at https://github.com/tektoncd/dashboard/pull/2345/commits/516a7838313823ea0df0362aacc2ec6fb1d850ce and https://github.com/tektoncd/dashboard/pull/2345/commits/d23ba12e70211f758cec62514f41d15e563a07f5 respectively. 

Please let me know if there are any corrections needed : ) 

Thank you for the opportunity to contribute this issue ☺️

File: src/api/pipelineRuns.js [hunk 6]
```diff
@@ -92,20 +93,20 @@ export function createPipelineRun({
...
     },
     spec: {
       pipelineRef: {
         name: pipelineName
       },
-      resources: Object.keys(resources).map(name => ({
-        name,
-        resourceRef: { name: resources[name] }
+      resources: Object.keys(resources).map(resourceName => ({
+        resourceName,
```
Comment (AlanGreene) [inline src/api/pipelineRuns.js:104]: This field should be called `name` to match the expected key in the spec. Unit tests related to this functionality are failing.

```suggestion
        name: resourceName,
```

File: src/api/pipelineRuns.js [hunk 7]
```diff
@@ -92,20 +93,20 @@ export function createPipelineRun({
...
+        resourceRef: { name: resources[resourceName] }
       })),
-      params: Object.keys(params).map(name => ({
-        name,
-        value: params[name]
+      params: Object.keys(params).map(paramName => ({
+        paramName,
```
Comment (AlanGreene) [inline src/api/pipelineRuns.js:108]: same as above, this should be `name`:

```suggestion
        name: paramName,
```

File: src/api/taskRuns.js [hunk 8]
```diff
@@ -110,9 +111,9 @@ export function createTaskRun({
     payload.spec.podTemplate = { nodeSelector };
   }
   if (params) {
-    payload.spec.params = Object.keys(params).map(name => ({
-      name,
-      value: params[name]
+    payload.spec.params = Object.keys(params).map(paramName => ({
+      paramName,
```
Comment (AlanGreene) [inline src/api/taskRuns.js:115]: Same as for the PipelineRun:

```suggestion
      name: paramName,
```

File: src/containers/CreatePipelineRun/CreatePipelineRun.js [hunk 9]
```diff
@@ -121,15 +123,18 @@ function CreatePipelineRun({ intl }) {
...
     },
     setState
   ] = useState({
     ...initialState,
     namespace: getNamespace(),
     pipelineRef: getPipelineName(),
     params: initialParamsState(null),
-    resources: initialResourcesState(null)
+    resources: initialResourcesState(null),
+    pipelineRunName: ''
```
Comment (AlanGreene) [inline src/containers/CreatePipelineRun/CreatePipelineRun.js:137]: This is already included in `initialState`, no need to repeat it here.

**Review by AlanGreene — CHANGES_REQUESTED** [review]
There are a few bugs related to params + resources, see comments below for suggested fixes.

You should be able to run the unit tests locally with `npm test -- --watchAll` to reproduce the failures and it will automatically rerun when you make changes.

Once you're done please also squash your commits down to one so we can merge it. Thanks 😸

**Comment by taechonkim** [comment]
> There are a few bugs related to params + resources, see comments below for suggested fixes.
> 
> You should be able to run the unit tests locally with `npm test -- --watchAll` to reproduce the failures and it will automatically rerun when you make changes.
> 
> Once you're done please also squash your commits down to one so we can merge it. Thanks 😸

@AlanGreene Sorry for late reply 😭
I fixed a part where an unit test error occured and confirmed that the unit tests are successful. 
Also squased all commits. 
Thank you for your help :D

**Comment by taechonkim** [comment]
> It looks like it's treating the new name field as required, I cannot create a PipelineRun or TaskRun if I leave the name field empty:
> 
> image

@AlanGreene Sorry :( I just fixed the bug and tested it. Please check again :) thank you

**Comment by AlanGreene** [comment]
Looks like it's still not working but it's a different error this time. I suspect you're passing an empty string as the value for `name` when the field is empty which means the default value at the API layer isn't being used, as default params only work when the value is `undefined`.

In CreateTaskRun / CreatePipelineRun when the user submits the form you could detect that the name is blank and replace the value with `undefined`. This should then work as expected.

image

**Comment by AlanGreene** [comment]
Opened a new PR to fix the bug above and update the unit tests: https://github.com/tektoncd/dashboard/pull/2383

**Pull request closed without merging**